VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp

```diff
@@ -0,0 +1,332 @@
+	{
+		float Value;
+		Material->GetScalarParameterValue(Info, Value);
+		NewMaterial->SetScalarParameterValueEditorOnly(Info, Value);
+	}
+
+	Material->GetAllTextureParameterInfo(ParameterInfos, ParameterIds);
+	for (const FMaterialParameterInfo& Info : ParameterInfos)
+	{
+		UTexture* Value;
+		Material->GetTextureParameterValue(Info, Value);
+		if (Value)
+		{
+			UTexture2D* PersistedTexture = SaveTexture(Cast<UTexture2D>(Value), Path, TextureCache);
+			NewMaterial->SetTextureParameterValueEditorOnly(Info, PersistedTexture);
+		}
+	}
+
+	Material->GetAllVectorParameterInfo(ParameterInfos, ParameterIds);
+	for (const FMaterialParameterInfo& Info : ParameterInfos)
+	{
+		FLinearColor Value;
+		Material->GetVectorParameterValue(Info, Value);
+		NewMaterial->SetVectorParameterValueEditorOnly(Info, Value);
+	}
+
+	MaterialCache.Add(Material, NewMaterial);
+
+	NewMaterial->PostEditChange();
+	MaterialPackage->MarkPackageDirty();
+
+	return NewMaterial;
+}
+
+UStaticMesh* SaveStaticMesh(UStaticMesh* Mesh, const FString& Path, FStaticMeshCache& MeshCache, FMaterialCache& MaterialCache,
+							FTextureCache& TextureCache)
+{
+	if (MeshCache.Contains(Mesh))
+	{
+		return MeshCache[Mesh];
+	}
+
+	// Create new StaticMesh Asset
+	const FString StaticMeshName = Mesh->GetName();
+	const FString PackageName = FPaths::Combine(Path, TEXT("Geometry"), StaticMeshName);
+	UPackage* MeshPackage = CreatePackage(*PackageName);
+	UStaticMesh* PersistedMesh = NewObject<UStaticMesh>(MeshPackage, *StaticMeshName, RF_Public | RF_Standalone);
+	PersistedMesh->InitResources();
+
+	FMeshDescription* OriginalMeshDescription = Mesh->GetMeshDescription(0);
+	FStaticMeshAttributes MeshAttributes(*OriginalMeshDescription);
+
+	// Copy Materials
+	const auto PolygonGroups = OriginalMeshDescription->PolygonGroups();
+	for (const auto& PolygonGroupId : PolygonGroups.GetElementIDs())
+	{
+		const FName MaterialName = MeshAttributes.GetPolygonGroupMaterialSlotNames()[PolygonGroupId];
+		int32 Index = Mesh->GetMaterialIndex(MaterialName);
+
+		if (Index != INDEX_NONE)
+		{
+			UMaterialInterface* Material = Mesh->GetMaterial(Index);
+
+			UMaterialInstanceDynamic* DynamicMaterial = Cast<UMaterialInstanceDynamic>(Material);
+			check(DynamicMaterial);
+			UMaterialInstanceConstant* NewMaterial = SaveMaterial(DynamicMaterial, Path, MaterialCache, TextureCache);
+			FName NewSlot = PersistedMesh->AddMaterial(NewMaterial);
+
+			MeshAttributes.GetPolygonGroupMaterialSlotNames()[PolygonGroupId] = NewSlot;
+		}
+	}
+
+	// Build the Static Mesh
+	TArray<const FMeshDescription*> MeshDescriptions;
+	MeshDescriptions.Add(OriginalMeshDescription);
+	PersistedMesh->BuildFromMeshDescriptions(MeshDescriptions);
+	PersistedMesh->PostEditChange();
+	PersistedMesh->MarkPackageDirty();
+
+	// Notify asset registry of new asset
+	FAssetRegistryModule::AssetCreated(PersistedMesh);
+
+	MeshCache.Add(Mesh, PersistedMesh);
+	return PersistedMesh;
+}
+} // namespace

```

Comment:
Same here but the textures/materials would need to be replaced after the duplicate call.